examples/hetero/bipartite_sage.py

```diff
@@ -0,0 +1,262 @@
+import argparse
+import os.path as osp
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from torch.nn import Embedding, Linear
+
+import torch_geometric.transforms as T
+from torch_geometric.datasets import MovieLens
+from torch_geometric.nn import Embedding, SAGEConv
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--use_weighted_loss', action='store_true',
+                    help='Whether to use weighted MSE loss.')
+args = parser.parse_args()
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+
+path = osp.join(osp.dirname(osp.realpath(__file__)), '../../data/MovieLens')
+user_sim_dict_path = 'swing_user_sim_dict.json'
+dataset = MovieLens(path, model_name='all-MiniLM-L6-v2')
+data = dataset[0].to(device)
+
+data['user'].x = torch.LongTensor(torch.arange(
+    0, data['user'].num_nodes)).to(device)
+del data['user'].num_nodes
+
+# Add a reverse ('movie', 'rev_rates', 'user') relation for message passing:
+data = T.ToUndirected()(data)
+del data['movie', 'rev_rates', 'user'].edge_label  # Remove "reverse" label.
+
+# Perform a link-level split into training, validation, and test edges:
+train_data, val_data, test_data = T.RandomLinkSplit(
+    num_val=0.1,
+    num_test=0.1,
+    neg_sampling_ratio=0.0,
+    edge_types=[('user', 'rates', 'movie')],
+    rev_edge_types=[('movie', 'rev_rates', 'user')],
+)(data)
+
+
+def to_u2i_mat(edge_index, i_num, u_num):
+    # Convert bipartite edge_index format to matrix format
+    u2imat = torch.zeros((u_num, i_num))
+    for i in range(edge_index.shape[1]):
+        u2imat[edge_index[0][i], edge_index[1][i]] = 1
+
+    return u2imat
+
+
+def get_coocur_mat(train_mat, num_neighbors):
+    # Generate the co-occurrence matrix with normalisation and top-k filtering
+    A = train_mat.T @ train_mat
+    n_items = A.shape[0]
+    res_mat = torch.zeros((n_items, num_neighbors))
+    res_sim_mat = torch.zeros((n_items, num_neighbors))
+
+    items_D = torch.sum(A, axis=0).reshape(-1)
+    users_D = torch.sum(A, axis=1).reshape(-1)
+    beta_uD = (1 / torch.sqrt(users_D + 1)).reshape(-1, 1)
+    beta_iD = (1 / torch.sqrt(items_D + 1)).reshape(1, -1)
+    all_ii_constraint_mat = beta_uD @ beta_iD
+
+    for i in range(n_items):
```

Comment:
Is there a need for this for-loop here? I assume you can apply `topk` in `dim=1` for this as well.

Reply:
Good suggestion, thx.